dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/dispatch/host/CommonHostManager.java

```diff
@@ -45,53 +51,96 @@
     protected ZookeeperRegistryCenter registryCenter;
 
     /**
-     * zookeeperNodeManager
+     * zookeeper node manager
      */
     @Autowired
     protected ZookeeperNodeManager zookeeperNodeManager;
 
+    /**
+     * zk master client
+     */
+    @Autowired
+    protected ZKMasterClient zkMasterClient;
+
+    /**
+     * worker group mapper
+     */
+    @Autowired
+    protected WorkerGroupMapper workerGroupMapper;
+
     /**
      * select host
      * @param context context
      * @return host
      */
     @Override
     public Host select(ExecutionContext context) {
-        Host host = new Host();
-        Collection<String> nodes = null;
+        List<HostWorker> candidates = null;
         String workerGroup = context.getWorkerGroup();
-        // executor type
         ExecutorType executorType = context.getExecutorType();
         switch (executorType) {
             case WORKER:
-                nodes = zookeeperNodeManager.getWorkerGroupNodes(workerGroup);
+                candidates = getHostWorkersFromDatabase(workerGroup);
```

Comment:
Need to call a database query for every command scheduling? This will cause the database to be very stressful, and you need to think about other ways.

Reply:
Isn’t it stressful to access zk every time? The performance of mysql is enough, don't worry at all

Reply:
I will consider optimizing database query, but the performance of database is still enough.

Reply:
By the way, if the database query causes stress, it seems that reading the data of several zk nodes maybe cause more stresses every time.

![image](https://user-images.githubusercontent.com/4902714/111936922-a5556f80-8b01-11eb-8e04-2d0a3b3973d2.png)

Reply:
Zk reads the local cache.

Reply:
Not at all @CalvinKirs